src/ray/raylet/node_manager.cc

```diff
@@ -1655,37 +1655,47 @@ void NodeManager::HandleRequestWorkerLease(const rpc::RequestWorkerLeaseRequest
-        // with normal task resource usages so GCS can update
-        // its resource view of this raylet.
         if (reply->rejected() && is_actor_creation_task) {
-          ResourceSet normal_task_resources =
-              local_task_manager_->CalcNormalTaskResources();
-          RAY_LOG(DEBUG) << "Reject leasing as the raylet has no enough resources."
-                         << " actor_id = " << actor_id << ", normal_task_resources = "
-                         << normal_task_resources.ToString() << ", local_resoruce_view = "
-                         << cluster_resource_scheduler_->GetClusterResourceManager()
-                                .GetNodeResourceViewString(
-                                    scheduling::NodeID(self_node_id_.Binary()));
           auto resources_data = reply->mutable_resources_data();
           resources_data->set_node_id(self_node_id_.Binary());
-          resources_data->set_resources_normal_task_changed(true);
-          auto &normal_task_map = *(resources_data->mutable_resources_normal_task());
-          normal_task_map = {normal_task_resources.GetResourceMap().begin(),
-                             normal_task_resources.GetResourceMap().end()};
-          resources_data->set_resources_normal_task_timestamp(
-              absl::GetCurrentTimeNanos());
+          // If resources are not enough due to normal tasks' preemption
+          // for GCS based actor scheduling, return
+          // with normal task resource usages so GCS can fast update
+          // its resource view of this raylet.
+          if (RayConfig::instance().gcs_actor_scheduling_enabled()) {
+            ResourceSet normal_task_resources =
+                local_task_manager_->CalcNormalTaskResources();
+            RAY_LOG(DEBUG) << "Reject leasing as the raylet has no enough resources."
+                           << " actor_id = " << actor_id << ", normal_task_resources = "
+                           << normal_task_resources.ToString()
+                           << ", local_resoruce_view = "
+                           << cluster_resource_scheduler_->GetClusterResourceManager()
+                                  .GetNodeResourceViewString(
+                                      scheduling::NodeID(self_node_id_.Binary()));
+            resources_data->set_resources_normal_task_changed(true);
+            auto &normal_task_map = *(resources_data->mutable_resources_normal_task());
+            normal_task_map = {normal_task_resources.GetResourceMap().begin(),
+                               normal_task_resources.GetResourceMap().end()};
+            resources_data->set_resources_normal_task_timestamp(
+                absl::GetCurrentTimeNanos());
+          }
         }
-
         send_reply_callback(status, success, failure);
       };
 
-  cluster_task_manager_->QueueAndScheduleTask(task,
+  if (request.grant_or_reject() && local_task_manager_) {
+    local_task_manager_->QueueAndScheduleTask(task,
```

Comment:
`grant_or_reject` == true means the request has already been scheduled, so it should bypass the `cluster_task_manager` and directly go to the local one. I think we can at least change it like:
 `  if (RayConfig::instance().gcs_actor_scheduling_enabled() && local_task_manager_) {
    local_task_manager_->QueueAndScheduleTask(task,`
I'm pretty sure gcs scheduler needs this, while causing no issues. For the other case with Raylet scheduling, we could double check. @scv119

Reply:
hmm this might not trivial to test. to really merge this behavior change we need to run the fullset of scheduler related tests (even nightly test) to ensure there is no regression. 

let's decouple that and do protocol change separately.

Reply:
I see.